File: python/src/tradingflow/operators/stocks/annualize.py
"""Annualize operator for YTD financial report data."""

from __future__ import annotations

import numpy as np

from ...operator import NativeOperator
from ...types import Handle, NodeKind


class Annualize(NativeOperator):
    """Convert year-to-date financial values into annualized values.

    Takes an input array of shape `(2 + N,)` laid out as
    `[year, day_of_year, ytd_1, ..., ytd_N]` (produced by
    [`FinancialReportSource`][tradingflow.sources.FinancialReportSource] with
    `with_report_date=True`) and outputs an array of shape `(N,)`
    containing annualized values.

    The operator handles any reporting frequency uniformly by using
    days-based scaling: `annualized = (current - last) × 365 / days_elapsed`.

    Parameters
    ----------
    source
        Handle to a `FinancialReportSource` output with
        `with_report_date=True`.
    """

    def __init__(self, source: Handle) -> None:
        # Input shape is (2 + N,); output shape is (N,).
        input_stride = 1
        for d in source.shape:
            input_stride *= d
        n = input_stride - 2
        if n < 1:
            raise ValueError(
                f"Annualize: input shape {source.shape} must have at least 3 elements "
                f"([year, day_of_year, ...values])"
            )
        output_shape = (n,) if n > 1 else ()

        super().__init__(
            native_id="annualize",
            inputs=(source,),
            kind=NodeKind.ARRAY,
            dtype=np.float64,
            shape=output_shape,
        )

File: crates/tradingflow/src/operators/stocks/annualize.rs
use crate::data::{Array, ArrayView, Instant};
use crate::graph::typed::Operator;
use crate::ports::ArrayPort;

/// Convert a YTD cumulative vector `[year, day_of_year, ytd_1..ytd_N]` into
/// annualized `[N]` values via days-based scaling (rank-1 in / rank-1 out).
#[derive(Clone, Default)]
pub struct Annualize;

impl Annualize {
    pub fn new() -> Self {
        Self
    }
}

/// Runtime state for [`Annualize`]: the previous-tick YTD snapshot plus the
/// output buffer.
pub struct AnnualizeState {
    prev_ytd: Vec<f64>,
    prev_year: i64,
    prev_day: f64,
    initialized: bool,
    out: Array<f64, 1>,
}

impl Operator for Annualize {
    type Inputs = ArrayPort<f64, 1>;
    type Outputs = ArrayPort<f64, 1>;
    type Context = Instant;
    type State = AnnualizeState;

    fn init(self, (_, view): (bool, ArrayView<'_, f64, 1>)) -> AnnualizeState {
        // Only the build-time input's shape is read here, to size the buffers.
        let input_len = view.to_contiguous().len();
        assert!(
            input_len >= 3,
            "Annualize: input must have shape [2 + N] with N >= 1, got length {input_len}"
        );
        let n = input_len - 2;
        AnnualizeState {
            prev_ytd: vec![0.0; n],
            prev_year: 0,
            prev_day: 0.0,
            initialized: false,
            out: Array::zeros([n]),
        }
    }

    fn compute<'a, 'b: 'a>(
        (_, view): (bool, ArrayView<'a, f64, 1>),
        state: &'b mut AnnualizeState,
        _: &Instant,
    ) -> (bool, ArrayView<'a, f64, 1>) {
        let input = view.to_contiguous();
        let year = input[0].floor() as i64;
        let day = input[1];
        let ytd = &input[2..];
        let n = ytd.len();
        let out = state.out.data_mut();

        let (is_new_year, days_elapsed) = if !state.initialized || year != state.prev_year {
            (true, day)
        } else {
            (false, day - state.prev_day)
        };

        if days_elapsed <= 0.0 {
            out.fill(f64::NAN);
        } else {
            let scale = 365.0 / days_elapsed;
            for i in 0..n {
                let delta = if is_new_year {
                    ytd[i]
                } else {
                    ytd[i] - state.prev_ytd[i]
                };
                out[i] = delta * scale;
            }
        }

        state.prev_ytd.copy_from_slice(ytd);
        state.prev_year = year;
        state.prev_day = day;
        state.initialized = true;
        (true, state.out.view())
    }

    fn passthrough<'a, 'b: 'a>(
        _: (bool, ArrayView<'a, f64, 1>),
        state: &'b mut AnnualizeState,
    ) -> (bool, ArrayView<'a, f64, 1>) {
        (false, state.out.view())
    }
}

/// Annualize a periodic (report-cadence) flow.
pub fn annualize() -> Annualize {
    Annualize::new()
}
